src/cdk/tree/tree.ts

```diff
@@ -398,21 +1256,58 @@ export class CdkTreeNode<T, K = T> implements FocusableOption, OnDestroy, OnInit
     this._destroyed.complete();
   }
 
-  /** Focuses the menu item. Implements for FocusableOption. */
+  getParent(): CdkTreeNode<T, K> | null {
+    return this._tree._getNodeParent(this) ?? null;
+  }
+
+  getChildren(): CdkTreeNode<T, K>[] | Observable<CdkTreeNode<T, K>[]> {
+    return this._tree._getNodeChildren(this);
+  }
+
+  /** Focuses this data node. Implemented for TreeKeyManagerItem. */
   focus(): void {
     this._elementRef.nativeElement.focus();
   }
 
-  // TODO: role should eventually just be set in the component host
-  protected _setRoleFromData(): void {
-    if (
-      !this._tree.treeControl.isExpandable &&
-      !this._tree.treeControl.getChildren &&
-      (typeof ngDevMode === 'undefined' || ngDevMode)
-    ) {
-      throw getTreeControlFunctionsMissingError();
+  /** Emits an activation event. Implemented for TreeKeyManagerItem. */
+  activate(): void {
+    if (this.isDisabled) {
+      return;
     }
-    this.role = 'treeitem';
+    this.activation.next(this._data);
+  }
+
+  /** Collapses this data node. Implemented for TreeKeyManagerItem. */
+  collapse(): void {
+    if (this.isExpandable) {
+      this._tree.collapse(this._data);
+    }
+  }
+
+  /** Expands this data node. Implemented for TreeKeyManagerItem. */
+  expand(): void {
+    if (this.isExpandable) {
+      this._tree.expand(this._data);
+    }
+  }
+
+  _setTabFocusable() {
+    this._elementRef.nativeElement.setAttribute('tabindex', '0');
+  }
+
+  _setTabUnfocusable() {
+    this._elementRef.nativeElement.setAttribute('tabindex', '-1');
+  }
+
+  _focusItem() {
+    if (this.isDisabled) {
+      return;
+    }
+    this._tree._keyManager.focusItem(this);
+  }
+
+  _emitExpansionState(expanded: boolean) {
+    this.expandedChange.emit(expanded);
```

Comment:
I would need to spend more time with this to make more specific recommendations, but I think there are some opportunities to refactor some logic out of this component class directly into other files

Reply:
Cassandra and I have looked into refactoring the rendering pipeline for the Tree, and we don't have a solution at this time. I don't think doing a big refactor of the tree's rendering pipeline is in scope for this PR, but maybe there is an incremental improvement we could to. The rendering pipeline is complicated :).

Would love to pick your brain on this, and see if there is something we can do to incrementally make this better.